contracts/Delegations.sol

```diff
@@ -30,66 +36,104 @@ contract Delegations is IDelegations, IStakeChangeNotifier, ContractRegistryAcce
-		require(to != prevDelegate, "delegation already in place");
 
-		uint256 prevStakePrevDelegate = uncappedStakes[prevDelegate];
-		uint256 prevStakeNewDelegate = uncappedStakes[to];
+		DelegateStatus memory prevDelegateStatusBefore = getDelegateStatus(prevDelegate);
+		DelegateStatus memory newDelegateStatusBefore = getDelegateStatus(to);
+
+		stakeOwnersData[from].delegation = to;
+
+		uint256 delegatorStake = getStakingContract().getStakeBalanceOf(from);
 
-		bool prevSelfDelegatingPrevDelegate = _isSelfDelegating(prevDelegate);
-		bool prevSelfDelegatingNewDelegate = _isSelfDelegating(to);
+		uncappedStakes[prevDelegate] = prevDelegateStatusBefore.uncappedStakes.sub(delegatorStake);
+		uncappedStakes[to] = newDelegateStatusBefore.uncappedStakes.add(delegatorStake);
 
-		delegations[msg.sender] = to;
+		DelegateStatus memory prevDelegateStatusAfter = getDelegateStatus(prevDelegate);
+		DelegateStatus memory newDelegateStatusAfter = getDelegateStatus(to);
 
-		uint256 delegatorStake = getStakingContract().getStakeBalanceOf(msg.sender);
+		totalDelegatedStake = totalDelegatedStake.sub(
+			prevDelegateStatusBefore.delegatedStake
+		).add(
+			prevDelegateStatusAfter.delegatedStake
+		).sub(
+			newDelegateStatusBefore.delegatedStake
+		).add(
+			newDelegateStatusAfter.delegatedStake
+		);
+
+		if (notifyElections) {
+			IElections elections = getElectionsContract();
+			IStakingContract staking = getStakingContract();
+
+			elections.delegatedStakeChange(
+				prevDelegate,
+				staking.getStakeBalanceOf(prevDelegate),
+				prevDelegateStatusAfter.delegatedStake
+			);
+
+			elections.delegatedStakeChange(
+				to,
+				staking.getStakeBalanceOf(to),
+				newDelegateStatusAfter.delegatedStake
+			);
+		}
 
-		uint256 newStakePrevDelegate = prevStakePrevDelegate.sub(delegatorStake);
-		uncappedStakes[prevDelegate] = newStakePrevDelegate;
+		emit Delegated(from, to);
 
-		uint256 newStakeNewDelegate = prevStakeNewDelegate.add(delegatorStake);
-		uncappedStakes[to] = newStakeNewDelegate;
+		if (delegatorStake != 0 && prevDelegate != to) {
+			emitDelegatedStakeChanged(prevDelegate, from, 0);
+			emitDelegatedStakeChanged(to, from, delegatorStake);
+		}
+	}
 
-		uint _totalDelegatedStake = totalDelegatedStake;
+	function delegate(address to) external onlyWhenActive {
+		delegateFrom(msg.sender, to, true);
+	}
 
-		_totalDelegatedStake = notifyOnDelegationChange(prevDelegate, prevSelfDelegatingPrevDelegate, prevStakePrevDelegate, _isSelfDelegating(prevDelegate), newStakePrevDelegate, _totalDelegatedStake);
-		_totalDelegatedStake = notifyOnDelegationChange(to, prevSelfDelegatingNewDelegate, prevStakeNewDelegate, _isSelfDelegating(to), newStakeNewDelegate, _totalDelegatedStake);
+	bool public delegationImportFinalized;
 
-		emit Delegated(msg.sender, to);
+	function importDelegations(address[] calldata from, address[] calldata to, bool notifyElections) external onlyWhenActive onlyMigrationOwner {
```

Comment:
Both: importDelegations and finalizeDelegationImport shouldn't require onlyWhenActive.
(In general onlyMigrationOwner should not depend on onlyWhenActive)
Add a modifier: onlyDuringDelegationImport